File: packages/python-sdk/strait/composition/_wait.py
"""Wait for run to reach terminal status with exponential backoff polling."""

from __future__ import annotations

import asyncio
import time
from dataclasses import dataclass
from typing import Any, Callable, TypeVar

from strait._errors import StraitTimeoutError

T = TypeVar("T")

_DEFAULT_TERMINAL_STATUSES = frozenset({
    "completed", "failed", "timed_out", "crashed",
    "system_failed", "canceled", "expired", "dead_letter",
})


@dataclass
class WaitForRunOptions:
    timeout_ms: int = 60_000
    initial_delay_ms: int = 500
    max_delay_ms: int = 10_000
    factor: float = 1.5
    is_terminal: Callable[[str], bool] | None = None


def wait_for_run(
    get_run: Callable[[str], T],
    get_status: Callable[[T], str],
    run_id: str,
    opts: WaitForRunOptions | None = None,
) -> T:
    o = opts or WaitForRunOptions()
    delay_ms = o.initial_delay_ms
    started_at = time.monotonic()

    while True:
        run = get_run(run_id)
        status = get_status(run)

        is_terminal = (
            o.is_terminal(status)
            if o.is_terminal is not None
            else status in _DEFAULT_TERMINAL_STATUSES
        )

        if is_terminal:
            return run

        elapsed = int((time.monotonic() - started_at) * 1000)
        if elapsed > o.timeout_ms:
            raise StraitTimeoutError(
                f"waitForRun timed out after {o.timeout_ms}ms for run {run_id}",
                run_id=run_id,
                elapsed_ms=elapsed,
            )

        time.sleep(delay_ms / 1000.0)
        delay_ms = int(min(o.max_delay_ms, max(1, round(delay_ms * o.factor))))


async def wait_for_run_async(
    get_run: Callable[[str], Any],
    get_status: Callable[[Any], str],
    run_id: str,
    opts: WaitForRunOptions | None = None,
) -> Any:
    o = opts or WaitForRunOptions()
    delay_ms = o.initial_delay_ms
    started_at = time.monotonic()

    while True:
        result = get_run(run_id)
        if asyncio.iscoroutine(result):
            run = await result
        else:
            run = result

        status = get_status(run)

        is_terminal = (
            o.is_terminal(status)
            if o.is_terminal is not None
            else status in _DEFAULT_TERMINAL_STATUSES
        )

        if is_terminal:
            return run

        elapsed = int((time.monotonic() - started_at) * 1000)
        if elapsed > o.timeout_ms:
            raise StraitTimeoutError(
                f"waitForRun timed out after {o.timeout_ms}ms for run {run_id}",
                run_id=run_id,
                elapsed_ms=elapsed,
            )

        await asyncio.sleep(delay_ms / 1000.0)
        delay_ms = int(min(o.max_delay_ms, max(1, round(delay_ms * o.factor))))

File: packages/go-sdk/composition/wait.go
package composition

import (
	"context"
	"fmt"
	"math"
	"time"

	strait "github.com/strait-dev/go-sdk"
)

// WaitForRunOptions configures polling behavior for WaitForRun.
type WaitForRunOptions struct {
	// TimeoutMs is the maximum wait time before timing out. Default 60000.
	TimeoutMs int
	// InitialDelayMs is the initial polling delay. Default 500.
	InitialDelayMs int
	// MaxDelayMs is the maximum polling delay. Default 10000.
	MaxDelayMs int
	// Factor is the exponential backoff multiplier. Default 1.5.
	Factor float64
	// IsTerminal overrides the default terminal status check.
	IsTerminal func(status string) bool
}

var defaultTerminalStatuses = map[string]bool{
	"completed":     true,
	"failed":        true,
	"timed_out":     true,
	"crashed":       true,
	"system_failed": true,
	"canceled":      true,
	"expired":       true,
	"dead_letter":   true,
}

// WaitForRun polls getRun until the run reaches a terminal status or times out.
func WaitForRun[T any](
	ctx context.Context,
	getRun func(ctx context.Context, runID string) (T, error),
	getStatus func(T) string,
	runID string,
	opts *WaitForRunOptions,
) (T, error) {
	var o WaitForRunOptions
	if opts != nil {
		o = *opts
	}
	if o.TimeoutMs <= 0 {
		o.TimeoutMs = 60000
	}
	if o.InitialDelayMs <= 0 {
		o.InitialDelayMs = 500
	}
	if o.MaxDelayMs <= 0 {
		o.MaxDelayMs = 10000
	}
	if o.Factor <= 0 {
		o.Factor = 1.5
	}

	var zero T
	delayMs := o.InitialDelayMs
	startedAt := time.Now()

	for {
		if err := ctx.Err(); err != nil {
			return zero, err
		}

		run, err := getRun(ctx, runID)
		if err != nil {
			return zero, err
		}

		status := getStatus(run)
		isTerminal := false
		if o.IsTerminal != nil {
			isTerminal = o.IsTerminal(status)
		} else {
			isTerminal = defaultTerminalStatuses[status]
		}

		if isTerminal {
			return run, nil
		}

		elapsed := time.Since(startedAt).Milliseconds()
		if elapsed > int64(o.TimeoutMs) {
			return zero, &strait.TimeoutError{
				Message:   fmt.Sprintf("waitForRun timed out after %dms for run %s", o.TimeoutMs, runID),
				RunID:     runID,
				ElapsedMs: elapsed,
			}
		}

		select {
		case <-ctx.Done():
			return zero, ctx.Err()
		case <-time.After(time.Duration(delayMs) * time.Millisecond):
		}

		delayMs = int(math.Min(float64(o.MaxDelayMs), math.Max(1, math.Round(float64(delayMs)*o.Factor))))
	}
}
